src/api/projectsAPI.ts

```diff
@@ -1,41 +1,143 @@
+  const projectInfo = await universalGet<ProjectGetResponse>(`/api/projects/${projectId}`);
+  const relativePaths = projectInfo.filepaths
+    .map((path) => path.replace(projectInfo.project_path + '/', '')) // Remove project path from API output
+    .map((path) => path.replace(/^\//, '')); // Remove the leading / (we don't want them in the output)
+
+  const rootFiles = relativePaths
+    .filter((path) => !path.startsWith('uploads'))
+    .map((path) => ({
+      path,
+      name: path.replace('/', ''),
+      children: undefined,
+    }));
+
+  const uploadsFiles = relativePaths
+    .filter((path) => path.startsWith('uploads/'))
+    .map((path) => ({
+      path,
+      name: path.replace('uploads/', ''),
+      children: undefined,
+    }));
+
+  return [
+    ...rootFiles,
+    {
+      path: 'uploads',
+      name: 'uploads',
+      children: uploadsFiles,
+    },
+  ];
+}
+
+// ########################################################################################
+// PROJECT FILE SYSTEM
+// ########################################################################################
+export const writeProjectTextFile = async (projectId: string, relativeFilePath: string, contents: string) => {
+  await universalPutFile(`/api/fs/${projectId}/${relativeFilePath}`, relativeFilePath, contents);
+};
+
+export const writeProjectBinaryFile = async (projectId: string, relativeFilePath: string, contents: ArrayBuffer) => {
+  await universalPutFile(`/api/fs/${projectId}/${relativeFilePath}`, relativeFilePath, contents);
+};
+
+export const readProjectTextFile = async (projectId: string, filePath: string) => {
+  try {
+    const responseBuffer = await universalGet<ArrayBuffer>(`/api/fs/${projectId}/${filePath}`, 'ArrayBuffer');
+    const textContent = new TextDecoder().decode(new Uint8Array(responseBuffer));
+    return textContent;
+  } catch (err) {
+    console.error('Error reading text file', err);
+    throw new Error('Invalid file format.');
+  }
+};
+
+export const readProjectBinaryFile = async (projectId: string, filePath: string) => {
+  try {
+    const responseBuffer = await universalGet<ArrayBuffer>(`/api/fs/${projectId}/${filePath}`, 'ArrayBuffer');
+    return new Uint8Array(responseBuffer);
+  } catch (err) {
+    console.error('Error reading binary file', err);
+    throw new Error('Invalid file format.');
+  }
+};
+
+export const existsProjectFile = async (projectId: string, filePath: string) => {
+  const status = await universalHead(`/api/fs/${projectId}/${filePath}`);
+  return status === 200;
+};
+
+export const deleteProjectFile = async (projectId: string, relativeFilePath: string) => {
+  await universalDelete(`/api/fs/${projectId}/${relativeFilePath}`);
+};
+
+export const renameProjectFile = async (projectId: string, relativeFilePath: string, newFilePath: string) => {
+  const fileExists = await existsProjectFile(projectId, relativeFilePath);
```

Comment:
Do we have an issue to implement a real rename endpoint?

Reply:
We discussed having that endpoint. But it won't fit into the filesystem API we designed. Maybe we can have a specific enpoint in the project API in the future, but not for now IMO.